File: backend/routes/items.js
const express = require('express');
const router = express.Router();
const bcrypt = require('bcryptjs');
const { v4: uuidv4 } = require('crypto').randomUUID ? { v4: () => require('crypto').randomUUID() } : require('crypto');
const Item = require('../models/Item');
const { verifyToken } = require('../middleware/auth');
const { upload } = require('../config/cloudinary');
const { runAutoMatching } = require('../utils/matching');

// Helper to generate a short readable secret code
const generateSecretCode = () => {
  const chars = 'ABCDEFGHJKLMNPQRSTUVWXYZ23456789';
  let code = '';
  for (let i = 0; i < 8; i++) {
    code += chars.charAt(Math.floor(Math.random() * chars.length));
  }
  return code;
};

// GET /api/items — Get all items (with optional filters)
router.get('/', async (req, res) => {
  try {
    const { type, category, status, page = 1, limit = 20 } = req.query;
    const filter = {};
    if (type) filter.type = type;
    if (category) filter.category = category;
    if (status) filter.status = status;
    else filter.status = 'active';

    const items = await Item.find(filter)
      .populate('reportedBy', 'name profileImage stars')
      .sort({ createdAt: -1 })
      .limit(Number(limit))
      .skip((Number(page) - 1) * Number(limit));

    const total = await Item.countDocuments(filter);

    res.json({ items, total, page: Number(page), pages: Math.ceil(total / limit) });
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
});

// GET /api/items/:id — Get single item
router.get('/:id', async (req, res) => {
  try {
    const item = await Item.findById(req.params.id).populate('reportedBy', 'name profileImage stars');
    if (!item) return res.status(404).json({ message: 'Item not found' });
    
    // Never expose secretCodeHash or secretCodePlain
    const itemObj = item.toObject();
    delete itemObj.secretCodeHash;
    delete itemObj.secretCodePlain;
    
    res.json(itemObj);
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
});

// POST /api/items — Report new item (lost or found)
router.post('/', verifyToken, upload.array('images', 5), async (req, res) => {
  try {
    const { type, title, description, category, location, date } = req.body;

    if (!type || !title || !description || !category || !location || !date) {
      return res.status(400).json({ message: 'All fields are required' });
    }

    const images = req.files ? req.files.map(f => f.path) : [];

    const itemData = {
      type,
      title,
      description,
      category,
      location,
      date: new Date(date),
      images,
      reportedBy: req.user._id,
      contactName: req.user.name,
      contactPhone: req.user.phone,
    };

    // For found items: generate secret code
    let plainCode;
    if (type === 'found') {
      plainCode = generateSecretCode();
      const salt = await bcrypt.genSalt(12);
      itemData.secretCodeHash = await bcrypt.hash(plainCode, salt);
    }

    const item = await Item.create(itemData);

    // Trigger auto-matching in background
    runAutoMatching(item).catch(console.error);

    const response = { message: 'Item reported successfully', item };
    if (type === 'found' && plainCode) {
      response.secretCode = plainCode;
      response.secretCodeNote = 'IMPORTANT: Save this code! Give it to the lost item owner to verify the handshake. It will not be shown again.';
    }

    res.status(201).json(response);
  } catch (err) {
    console.error('Create item error:', err);
    res.status(500).json({ message: err.message });
  }
});

// PATCH /api/items/:id — Update item (only owner)
router.patch('/:id', verifyToken, async (req, res) => {
  try {
    const item = await Item.findById(req.params.id);
    if (!item) return res.status(404).json({ message: 'Item not found' });
    if (item.reportedBy.toString() !== req.user._id.toString()) {
      return res.status(403).json({ message: 'Not authorized' });
    }

    const allowed = ['title', 'description', 'location', 'date'];
    allowed.forEach(f => { if (req.body[f] !== undefined) item[f] = req.body[f]; });
    await item.save();

    res.json({ message: 'Item updated', item });
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
});

// DELETE /api/items/:id — Delete item (only owner)
router.delete('/:id', verifyToken, async (req, res) => {
  try {
    const item = await Item.findById(req.params.id);
    if (!item) return res.status(404).json({ message: 'Item not found' });
    if (item.reportedBy.toString() !== req.user._id.toString()) {
      return res.status(403).json({ message: 'Not authorized' });
    }
    await item.deleteOne();
    res.json({ message: 'Item deleted' });
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
});

// GET /api/items/my/items — Get current user's items
router.get('/my/items', verifyToken, async (req, res) => {
  try {
    const items = await Item.find({ reportedBy: req.user._id }).sort({ createdAt: -1 });
    res.json(items);
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
});

module.exports = router;

File: backend/routers/items.py
from fastapi import APIRouter, Depends, HTTPException, Form, UploadFile, File
from typing import List, Optional
from config import db
from dependencies import get_current_user
import cloudinary.uploader
from bson import ObjectId
import datetime
from utils.matching import run_auto_matching

router = APIRouter()

@router.post("")
async def create_item(
    type: str = Form(...), title: str = Form(...), description: str = Form(...),
    category: str = Form(...), location: str = Form(...), date: str = Form(...),
    contactName: Optional[str] = Form(None), contactPhone: Optional[str] = Form(None),
    images: Optional[List[UploadFile]] = File(None),
    user=Depends(get_current_user)
):
    if images is None: images = []
    image_urls = []
    for img in images:
        if not img.filename: continue
        try:
            res = cloudinary.uploader.upload(img.file)
            image_urls.append(res.get("secure_url"))
        except Exception as e:
            print("Cloudinary error:", e)
            
    new_item = {
        "type": type, "title": title, "description": description,
        "category": category, "location": location, "date": datetime.datetime.fromisoformat(date.replace("Z", "+00:00")) if "T" in date else datetime.datetime.utcnow(),
        "contactName": contactName or user.get("name", ""), "contactPhone": contactPhone or user.get("phone", ""),
        "images": image_urls, "reportedBy": user["_id"], "status": "active",
        "createdAt": datetime.datetime.utcnow(), "updatedAt": datetime.datetime.utcnow()
    }
    
    res = await db.items.insert_one(new_item)
    new_item["_id"] = res.inserted_id
    
    # Run auto matching synchronously (Required for Vercel Serverless)
    await run_auto_matching(new_item)
    
    new_item["_id"] = str(new_item["_id"])
    new_item["reportedBy"] = str(new_item["reportedBy"])
    return new_item

@router.get("/my/items")
async def get_my_items(user=Depends(get_current_user)):
    cursor = db.items.find({"reportedBy": user["_id"]}).sort("createdAt", -1)
    items = []
    async for doc in cursor:
        doc["_id"] = str(doc["_id"])
        doc["reportedBy"] = str(doc["reportedBy"])
        items.append(doc)
    return items

@router.get("")
async def get_items(type: Optional[str] = None, user_id: Optional[str] = None, status: Optional[str] = "active"):
    query = {}
    if type: query["type"] = type
    if user_id: query["reportedBy"] = ObjectId(user_id)
    if status and status != "all": query["status"] = status
    
    cursor = db.items.find(query).sort("createdAt", -1)
    items = []
    async for doc in cursor:
        doc["_id"] = str(doc["_id"])
        doc["reportedBy"] = str(doc["reportedBy"])
        items.append(doc)
    return {"items": items}

@router.get("/{id}")
async def get_item(id: str):
    item = await db.items.find_one({"_id": ObjectId(id)})
    if not item: raise HTTPException(status_code=404, detail="Item not found")
    
    user_doc = await db.users.find_one({"_id": ObjectId(item["reportedBy"])})
    if user_doc:
        item["reportedBy"] = {
            "_id": str(user_doc["_id"]),
            "name": user_doc.get("name"),
            "profileImage": user_doc.get("profileImage"),
            "stars": user_doc.get("stars", 0)
        }
    else:
        item["reportedBy"] = str(item["reportedBy"])
        
    item["_id"] = str(item["_id"])
    return item

@router.delete("/{id}")
async def delete_item(id: str, user=Depends(get_current_user)):
    item = await db.items.find_one({"_id": ObjectId(id)})
    if not item: raise HTTPException(404, "Not found")
    if str(item["reportedBy"]) != str(user["_id"]): raise HTTPException(403, "Not authorized")
    
    # Delete from cloudinary
    if item.get("images"):
        for url in item["images"]:
            try:
                # Extract public_id roughly (assuming standard format)
                public_id = url.split("/")[-1].split(".")[0]
                cloudinary.uploader.destroy(public_id)
            except Exception as e:
                print("Cloudinary delete error:", e)
                
    # Delete from mongo
    await db.items.delete_one({"_id": ObjectId(id)})
    # Delete related matches
    await db.matches.delete_many({"$or": [{"lostItem": ObjectId(id)}, {"foundItem": ObjectId(id)}]})
    return {"message": "Item deleted successfully"}
